datafusion/common/src/error.rs

```diff
@@ -450,6 +396,99 @@ impl DataFusionError {
+        let error_name = self.get_error_name();
+        match *self {
+            DataFusionError::ArrowError(ref desc, ref backtrace) => {
+                let backtrace = backtrace.clone().unwrap_or("".to_owned());
+                write!(f, "{error_name}{desc}{backtrace}")
+            }
+            #[cfg(feature = "parquet")]
+            DataFusionError::ParquetError(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            #[cfg(feature = "avro")]
+            DataFusionError::AvroError(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::IoError(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::SQL(ref desc, ref backtrace) => {
+                let backtrace: String = backtrace.clone().unwrap_or("".to_owned());
+                write!(f, "{error_name}{desc:?}{backtrace}")
+            }
+            DataFusionError::Configuration(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::NotImplemented(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::Internal(ref desc) => {
+                write!(f, "{error_name}{desc}.\nThis was likely caused by a bug in DataFusion's \
+                    code and we would welcome that you file an bug report in our issue tracker")
+            }
+            DataFusionError::Plan(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::SchemaError(ref desc, ref backtrace) => {
+                let backtrace: &str =
+                    &backtrace.as_ref().clone().unwrap_or("".to_owned());
+                write!(f, "{error_name}{desc}{backtrace}")
+            }
+            DataFusionError::Execution(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::ResourcesExhausted(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::External(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            #[cfg(feature = "object_store")]
+            DataFusionError::ObjectStore(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+            DataFusionError::Context(ref desc, ref err) => {
+                write!(f, "{error_name}{desc}\ncaused by\n{}", *err)
+            }
+            DataFusionError::Substrait(ref desc) => {
+                write!(f, "{error_name}{desc}")
+            }
+        }
+    }
+
+    /// Strips error description out of the error message
+    /// Error has a format `{error_name}{`error`}
+    /// The method strips the `error_name` from level and outputs 'error``
+    pub fn strip_error_name(self) -> String {
+        self.to_string().replace(self.get_error_name(), "")
```

Comment:
I think it would be better if we could avoid including the error message in the first place rather than removing it later. There is the risk of accidentally removing text that was not added by `impl Display for DataFusionError`.

Reply:
I was thinking about this in the first place. 

- We can fix the formatter so it will skip the error name for some format options, that doesn't work with `to_string()` which has default formatter.

- we can tweak `impl Display` with cargo feature and return name as blank if the feature enabled and that looks little bit overcompicated.

- we can refer to env variable and depending on the value define the error name but this might be expensive. 

- we can probably introduce a DF config param?

Reply:
I am not sure exactly what the usecase is, but if seems like if the only need is:

> Sometimes it is required to fetch the original message without Datafusion error name.

Rather than trying to fix the message after it has been created, how about we add a function that just provides the underlying message? Something like

```rust
impl DataFusionError {
  /// retrieve the underlying error message for this error, without the
  /// type prefix. 
  ///
  /// For example, if err.to_string() returns `Error during planning: My Query Error` 
  /// this function would return only `My Query Error`
  fn message(&self) -> &str { .. }
}
```

And then the impl could be

```rust
impl Display for DataFusionError {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        match *self {
            DataFusionError::ArrowError(ref desc, ref backtrace) => {
                let backtrace = backtrace.clone().unwrap_or("".to_owned());
                write!(f, "Arrow error: {}{backtrace}", self.message())
            }
            #[cfg(feature = "parquet")]
            DataFusionError::ParquetError(ref desc) => {
                write!(f, "Parquet error: {}", message)
            }
...
}
```


Maybe ` fn message(&self)` would have to return `Cow<str>` or something to avoid copying in the external cases....

Reply:
This seems like the simplest solution.

Our use case (Comet) is that we are using [thiserror](https://crates.io/crates/thiserror) and we are delegating to DataFusionError for the error string and it currently uses the `Display` trait.

@comphead I don't know enough about thiserror yet to know if/how we could call the new method proposed here. Do you know?

Reply:
looks like it works even without thiserror magic, we can call `.message()` method in `From<>` error conversion